### ISSUE TITLE
Migrate NYT API example to a public API

### ISSUE THREAD

**Issue by jfc3-dol** [mdn/content#26602]
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Third_party_APIs

### What specific section or headline is this issue about?

Wiring up the pagination buttons

### What information was incorrect, unhelpful, or incomplete?

The New York Times API example page ( https://mdn.github.io/learning-area/javascript/apis/third-party-apis/nytimes/finished/ ) found at the link “see it running live here” under the heading “Wiring up the pagination buttons” does not show any results or even displays an error message that something is wrong.

**NOTE** - The same applies for the YouTube API example page ( https://mdn.github.io/learning-area/javascript/apis/third-party-apis/youtube/ ).

I'm using Chrome on my Windows 10 machine.

### What did you expect to see?

I expected to see a few articles at least for the search term "food".

### Do you have any supporting links, references, or citations?

no

### Do you have anything more you want to share?

no

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/javascript/client-side_web_apis/third_party_apis`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Third_party_APIs
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/javascript/client-side_web_apis/third_party_apis/index.md
* Last commit: https://github.com/mdn/content/commit/4def230f85756724b59660e3cd9de363db724ef8
* Document last modified: 2023-02-24T00:05:28.000Z

</details>

**Comment by kevinctofel** [comment]
@jfc3-dol I checked the code of the two examples and it looks like they have either missing or outdated API keys. However, if you follow the instructions to get an API key for both the NYTimes or YouTube and add it in the code, per the instructions, the examples do work. That's just a way to get around this for now, although it does teach you something! ;)

@wbamberg pinging you as the original contributor here; sorry if I should tag someone else. 

Using Developer Tools, I see the network request for the NYTimes example returns a 401 - not authorized response. I suspect the API key is no longer valid. Using my own API key, the code does work as expected. I doubt I can request one on behalf of Mozilla, but if I can, just let me know and I'll submit a PR to address this.

For the YouTube live example, the code doesn't actually have an API key at all. Ideally, a reader would follow the example to get a key and add it to the code but the documentation also [points to a "working" example](https://mdn.github.io/learning-area/javascript/apis/third-party-apis/youtube/). Again, I could submit a PR with an API key for the example but it would be my personal one, which isn't ideal. Cheers!

**Comment by jfc3-dol** [comment]
@kevinctofel 
Thanks, for pointing out that the text/instructions work as they should.
My concern is that the example pages for the New York Times and YouTube that MDN point to do not work to show a person what the results should look like.
I thought maybe the API version or as you mentioned the API key is no longer valid.

**Comment by kevinctofel** [comment]
> My concern is that the example pages for the New York Times and YouTube that MDN point to do not work to show a person what the results should look like. 

Completely agree with that valid concern. I'll submit a PR with a either a new, personal API key or see if I can get one connected to Mozilla. I may have to submit a second PR with the finished YouTube project code as well. Cheers!

**Comment by jfc3-dol** [comment]
@kevinctofel 

Splitting the two example pages into two issues is fine. I typically make them two, but I only put them together to save myself a bit of time and know better than to make them separate. 

Sorry about that.

Good luck getting it all figured out.

**Comment by wbamberg** [comment]
Thanks for filing, @jfc3-dol , and for the investigation, @kevinctofel !

The YT example is working as intended: [the example expects you to get your own key and plug it into the code](https://github.com/mdn/learning-area/blob/main/javascript/apis/third-party-apis/youtube/script.js#L19-L30).

But even just by linking to https://mdn.github.io/learning-area/javascript/apis/third-party-apis/youtube/, the page [does imply it will work](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Third_party_APIs#youtube_example), which it probably shouldn't.

For the NYT example: yes, it look as if the problem is that the key has expired. Actually, although I did refactor these pages a little while ago, I didn't get the key: I think that would have been @chrisdavidmills, [back in 2019](https://github.com/mdn/learning-area/commit/3c0e38b2952e48595aa25ee459afb80573f3f3ee#diff-47dd1c258a681f10cb6a94defd351b14b61049f17115f443e8c0fed0fbc53d80). I don't know if Chris used some official Mozilla account or his own personal account.

I can't personally use an official Mozilla account, even if one exists, and like you I don't think I should use my own.

I don't know if we have a policy for API keys in examples. I am uncomfortable with including live keys in public examples at all, and think we should probably do what the YouTube example here does and expect readers to get their own. However this obviously increases the barrier to checking out the example.

I'll ask around and see what people think.

**Comment by kevinctofel** [comment]
@wbamberg yup, I'm right there with you on not including live keys in public examples. I wonder if a better approach is to actually rework both of the examples using a .env file and API key variable? 

I'm sure part of the goal here is simplicity but... it would eliminate the issues of having an API key exposed publicly. And it would introduce the concept of how to keep and use API keys in a secure way. We could then expose new APIs on the server where the example apps run so they would work.

Either way, as a new dev-in-training, I'm happy to work up PRs for these once we have some direction; just let me know how the conversations go. Cheers!

**Comment by chrisdavidmills** [comment]
Using an .env would make sense.

When I first wrote that example, I seem to remember it being very hard to find an API that was free, but also useful and interesting, which represented what I wanted to show. I also wanted to show a live example, as it feels really lame to show the example but not have it actually working.

It would also be great to show a good practice example of how to use API keys securely; that kind of info is really lacking on MDN.

**Comment by kevinctofel** [comment]
Unless someone feels otherwise, I propose that I update the documentation and working YouTube app code with an .env file. The working code example would then work and readers would better understand how to hide their API keys. 

Obviously, I don't know how much traffic the code example implementation gets (or would get) but I suspect it's low enough that I can use my own API key for it. If anyone at Mozilla can share traffic data though, that would be useful to make the decision. I can do the same (in a separate PR) for the NYT code example with a personal API key for the same reason.

Ideally, a Mozilla-based user API key would be best so if that's still an option, I can use that. I'm just not sure who we need to contact.

**Comment by jfc3-dol** [comment]
@kevinctofel 

Could you create a new accounts for each with a different e-mail address? So not to use your personal one.

**Comment by kevinctofel** [comment]
Good idea @jfc3-dol! It will still be tied to me but at least it won't affect my main email address used for other work. Cheers!

**Comment by jfc3-dol** [comment]
@kevinctofel

Glad I could help with my suggestion.

**Comment by wbamberg** [comment]
@kevinctofel , thank you for the update. Actually I was looking at this last night. Do you know how to use a .env file with GitHub Pages? It wasn't at all obvious to me, but I might be missing something.

As far as the key itself goes, I don't think you should use a personal key. If we can work out the technical side, I will try to get an answer out of Mozilla about a key to use.

**Comment by Josh-Cena** [comment]
@wbamberg @kevinctofel Do any of you have a solid way to progress? Do you need any input from others?

**Comment by wbamberg** [comment]
> Do any of you have a solid way to progress?

The proposal in this issue is to get an API key from Mozilla and keep it in a secret that's accessible to GitHub pages, and update the documentation to explain this. I'm not sure it's possible to create a secret that is accessible in a GitHub Pages site, but https://github.com/mdn/content/issues/26602#issuecomment-1563172215 indicates that it is possible. So it takes someone to look into this and see how/if it can be done.

Personally, I think this page should separate key management out of the rest of "using a third-party API". It should:

- choose an API that doesn't require authentication (like say https://api.artic.edu/docs/#introduction) to cover the mechanics of how to use an API, with a finished live sample that actually works
- then choose an API that does require authentication, and talk properly about secret management, and walk through how to do it, probably without a finished live sample that actually works

Anyway, I think it's not so much about needing a decision as needing someone to put work into a solution.

**Comment by Josh-Cena** [comment]
GitHub pages is purely front end, so anything you include in the site is accessible in client side anyway and is not more secure than including it in the source code (other than possibly security by obscurity). I don't know how this API is supposed to work, but if you need your API connector to be on the backend so it can security transmit API keys to YouTube without exposing it to the client, then you need an actual server hosted somewhere. This goes for any API that requires auth you may have in mind.

For  the client toolchain example, I built https://mdn.github.io/client-toolchain-example/, which uses the GitHub API that is fully public. Perhaps you can consider that. See also: https://github.com/mdn/content/issues/1455

**Comment by wbamberg** [comment]
> GitHub pages is purely front end, so anything you include in the site is accessible in client side anyway and is not more secure than including it in the source code (other than possibly security by obscurity). I don't know how this API is supposed to work, but if you need your API connector to be on the backend so it can security transmit API keys to YouTube without exposing it to the client, then you need an actual server hosted somewhere. This goes for any API that requires auth you may have in mind.
> 

Yes, that was my understanding as well.
